src/client/Client.cc

```diff
@@ -10760,6 +10761,56 @@ int Client::_flock(Fh *fh, int cmd, uint64_t owner)
   return ret;
 }
 
+int Client::get_snap_info(const char *path, const UserPerm &perms, struct snap_info *snap_info) {
```

Comment:
This is the C++ interface so we can use real containers here inside of `struct snap_info` (which will actually simplify a lot of your code because you can encode/decode std::maps without NUL char gymnastics.) Only libcephfs.cc needs to have a C-friendly interface.

Reply:
`struct snap_info*` should be some `Client` defined struct for the `Client` interface. Something like:

````
struct Client::snap_info {
  uint64_t id;
  std::map<std::string,std::string> metadata;
}
````